File: experiments/source/python/mixed_book_signal_reassociated.py
import loom


@loom.entry
def mixed_book_signal_reassociated(
    scale: float, epsilon: float, threshold: float, bid: loom.Tensor1, ask: loom.Tensor1
) -> loom.Tensor1:
    def signal(bid_i: float, ask_i: float) -> float:
        depth = ask_i + bid_i
        spread = bid_i - ask_i
        normalized = spread / (epsilon + depth)
        return normalized * scale if threshold < depth else 0.0

    return loom.Tensor1.map2(signal, bid, ask)

File: experiments/source/ocaml/mixed_book_signal_reassociated.ml
open Loom

let[@loom.entry] mixed_book_signal_reassociated (scale : float) (epsilon : float)
    (threshold : float) (bid : Tensor1.t) (ask : Tensor1.t) =
  let signal bid_i ask_i =
    let depth = ask_i +. bid_i in
    let spread = bid_i -. ask_i in
    let normalized = spread /. (epsilon +. depth) in
    if threshold < depth then normalized *. scale else 0.0
  in
  Tensor1.map2 signal bid ask
